[Model element] Support loading textures in GPU process model element

mwyrzykowski opened the pull request:
#### ff311a2fc4654fd14cf023e5537e2d052e35face
<pre>
[Model element] Support loading textures in GPU process model element
<a href="https://bugs.webkit.org/show_bug.cgi?id=299470">https://bugs.webkit.org/show_bug.cgi?id=299470</a>
<a href="https://rdar.apple.com/161267992">rdar://161267992</a>

Reviewed by Etienne Segonzac.

Send materials and textures from the WCP to to GPU process for rendering.

Remove the WebKit:: types which overlapped with WebCore:: types. Previously
we were converted back and forth from otherwise identical WebKit and WebCore
types which seemed redundant and also inefficient.

Test: model-element/gpup-model-element-textures.html
* LayoutTests/model-element/gpup-model-element-textures-expected.png: Added.
* LayoutTests/model-element/gpup-model-element-textures-expected.txt: Added.
* LayoutTests/model-element/gpup-model-element-textures.html: Added.
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.cpp:
(WebCore::DDModel::DDMeshImpl::addMesh):
(WebCore::DDModel::DDMeshImpl::update):
(WebCore::DDModel::convert):
(WebCore::DDModel::DDMeshImpl::addTexture):
(WebCore::DDModel::DDMeshImpl::updateTexture):
(WebCore::DDModel::convertArray):
(WebCore::DDModel::DDMeshImpl::addMaterial):
(WebCore::DDModel::DDMeshImpl::updateMaterial):
* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.h:
* Source/WebCore/Modules/Model/InternalAPI/DDImageAsset.h: Renamed from Source/WebKit/Shared/Model/DDReplaceVertices.h.
* Source/WebCore/Modules/Model/InternalAPI/DDMaterialDescriptor.h: Added.
* Source/WebCore/Modules/Model/InternalAPI/DDMesh.h:
* Source/WebCore/Modules/Model/InternalAPI/DDMeshDescriptor.h:
* Source/WebCore/Modules/Model/InternalAPI/DDMeshPart.h:
* Source/WebCore/Modules/Model/InternalAPI/DDModel.serialization.in:
* Source/WebCore/Modules/Model/InternalAPI/DDReplaceVertices.h:
* Source/WebCore/Modules/Model/InternalAPI/DDTextureDescriptor.h: Renamed from Source/WebKit/Shared/Model/DDVertexLayout.h.
* Source/WebCore/Modules/Model/InternalAPI/DDUpdateMaterialDescriptor.h: Copied from Source/WebCore/Modules/Model/InternalAPI/DDReplaceVertices.h.
* Source/WebCore/Modules/Model/InternalAPI/DDUpdateMeshDescriptor.h:
* Source/WebCore/Modules/Model/InternalAPI/DDUpdateTextureDescriptor.h: Copied from Source/WebCore/Modules/Model/InternalAPI/DDReplaceVertices.h.
* Source/WebCore/Modules/Model/InternalAPI/DDVertexAttributeFormat.h:
* Source/WebCore/Modules/Model/InternalAPI/DDVertexLayout.h:
* Source/WebCore/Modules/Model/ModelDDInlineConverters.h:
(WebCore::toCpp):
(WebCore::toVector):
* Source/WebCore/Modules/Model/ModelDDTypes.h:
* Source/WebCore/Modules/model-element/DDModelPlayer.h:
* Source/WebCore/Modules/model-element/DDModelPlayer.mm:
(WebCore::DDModelPlayer::load):
(WebCore::DDModelPlayer::update):
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
(): Deleted.
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayer.cpp:
(WebCore::DummyModelPlayer::load):
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayer.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/canvas/UsdModelLoader.swift:
(update(_:)):
(meshCreated(_:)):
(meshUpdated(_:data:)):
(meshDestroyed(_:)):
(materialCreated(_:)):
(materialUpdated(_:data:)):
(materialDestroyed(_:)):
(textureCreated(_:)):
(textureUpdated(_:data:)):
(textureDestroyed(_:)):
(Converter.webUpdateMeshRequestFromUpdateMeshRequest(_:)):
(Converter.toWebSemantic(_:)):
(Converter.toWebImageAsset(_:id:)):
(Converter.toWebDDNodeType(_:)):
(Converter.toWebDDBuiltin(_:)):
(Converter.constantValues(_:WebDDConstant:)):
(Converter.toWebDDConstantContainer(_:)):
(Converter.toWebNodes(_:)):
(Converter.toWebEdges(_:)):
(Converter.toWebDDDataType(_:)):
(Converter.toWebInputOutputs(_:)):
(Converter.toWebPrimvars(_:)):
(Converter.toWebMaterialGraph(_:id:)):
(Converter.webAddMeshRequestFromAddMeshRequest(_:)):
(Converter.webUpdateTextureRequestFromUpdateTextureRequest(_:)):
(Converter.webAddTextureRequestFromAddTextureRequest(_:)):
(Converter.webUpdateMaterialRequestFromUpdateMaterialRequest(_:)):
(Converter.webAddMaterialRequestFromAddMaterialRequest(_:)):
(WebUSDModelLoader.textureAddedCallback):
(WebUSDModelLoader.textureUpdatedCallback):
(WebUSDModelLoader.materialAddedCallback):
(WebUSDModelLoader.materialUpdatedCallback):
(WebUSDModelLoader.retainedRequests):
(WebUSDModelLoader.update(_:)):
(WebUSDModelLoader.requestCompleted(_:)):
(WebUSDModelLoader.updateMesh(_:)):
(WebUSDModelLoader.addTexture(_:)):
(WebUSDModelLoader.updateTexture(_:)):
(WebUSDModelLoader.addMaterial(_:)):
(WebUSDModelLoader.updateMaterial(_:)):
(WebUSDModelLoader.meshAdded(_:)):
* Source/WebGPU/WebGPU/DDMesh.h:
* Source/WebGPU/WebGPU/DDMesh.mm:
(WebGPU::DDMesh::DDMesh):
(WebGPU::convert):
(WebGPU::convertDescriptor):
(WebGPU::DDMesh::update):
(WebGPU::DDMesh::addMesh):
(WebGPU::DDMesh::addTexture):
(WebGPU::DDMesh::updateTexture):
(WebGPU::DDMesh::addMaterial):
(WebGPU::DDMesh::updateMaterial):
(wgpuDDTextureUpdate):
(wgpuDDTextureAdd):
(wgpuDDMaterialUpdate):
(wgpuDDMaterialAdd):
* Source/WebGPU/WebGPU/DDModelTypes.h:
* Source/WebGPU/WebGPU/UsdModelRenderer.swift:
(NodeStore.nodeDefinitionStore):
(Helper.makeDirectTextureDescriptorFromMTLTexture(_:)):
(Helper.pixelFormat):
(sgDataTypeFromInputOutputDataType(_:)):
(nodes):
(allNodes):
(connections):
(customSurfaceShaderParameterBindings):
(assetPathToPropertyName):
(pbrInputNameToType(_:)):
(convert(_:)):
(convertToAddMeshRequest(_:id:)):
(DDBridgeReceiver.meshInstances):
(DDBridgeReceiver.meshResources):
(DDBridgeReceiver.textures):
(DDBridgeReceiver.textureFromPath):
(DDBridgeReceiver.render(with:)):
(DDBridgeReceiver.addMesh(_:identifier:)):
(DDBridgeReceiver.addTextureAsync(_:identifier:)):
(DDBridgeReceiver.addTexture(_:identifier:)):
(DDBridgeReceiver.updateTextureAsync(_:identifier:)):
(DDBridgeReceiver.updateTexture(_:identifier:)):
(DDBridgeReceiver.addMaterialAsync(_:identifier:)):
(DDBridgeReceiver.addMaterial(_:identifier:)):
(DDBridgeReceiver.updateMaterialAsync(_:identifier:)):
(DDBridgeReceiver.updateMaterial(_:identifier:)):
(DDBridgeReceiver.updateMeshAsync(_:identifier:)):
(DDBridgeReceiver.updateMesh(_:identifier:)):
(DDBridgeReceiver.device): Deleted.
(DDBridgeReceiver.context): Deleted.
(DDBridgeReceiver.scene): Deleted.
(DDBridgeReceiver.meshes): Deleted.
(DDBridgeReceiver.camera): Deleted.
(DDBridgeReceiver.commandQueue): Deleted.
(DDBridgeReceiver.captureManager): Deleted.
(DDBridgeReceiver.setDeviceWith(_:)): Deleted.
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.cpp:
(WebKit::RemoteDDMesh::addMesh):
(WebKit::RemoteDDMesh::update):
(WebKit::RemoteDDMesh::addTexture):
(WebKit::RemoteDDMesh::updateTexture):
(WebKit::RemoteDDMesh::addMaterial):
(WebKit::RemoteDDMesh::updateMaterial):
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.h:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/Model/DDMeshDescriptor.cpp: Removed.
* Source/WebKit/Shared/Model/DDMeshDescriptor.h: Removed.
* Source/WebKit/Shared/Model/DDMeshDescriptor.serialization.in: Removed.
* Source/WebKit/Shared/Model/DDMeshPart.cpp: Removed.
* Source/WebKit/Shared/Model/DDMeshPart.h: Removed.
* Source/WebKit/Shared/Model/DDMeshPart.serialization.in: Removed.
* Source/WebKit/Shared/Model/DDReplaceVertices.cpp: Removed.
* Source/WebKit/Shared/Model/DDReplaceVertices.serialization.in: Removed.
* Source/WebKit/Shared/Model/DDUpdateMeshDescriptor.cpp: Removed.
* Source/WebKit/Shared/Model/DDUpdateMeshDescriptor.h: Removed.
* Source/WebKit/Shared/Model/DDUpdateMeshDescriptor.serialization.in: Removed.
* Source/WebKit/Shared/Model/DDVertexAttributeFormat.h: Removed.
* Source/WebKit/Shared/Model/DDVertexAttributeFormat.serialization.in: Removed.
* Source/WebKit/Shared/Model/DDVertexLayout.serialization.in: Removed.
* Source/WebKit/Shared/Model/ModelConvertFromBackingContext.h:
* Source/WebKit/Shared/Model/ModelConvertToBackingContext.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.cpp:
(WebKit::DDModel::RemoteDDMeshProxy::addMesh):
(WebKit::DDModel::RemoteDDMeshProxy::update):
(WebKit::DDModel::RemoteDDMeshProxy::addTexture):
(WebKit::DDModel::RemoteDDMeshProxy::updateTexture):
(WebKit::DDModel::RemoteDDMeshProxy::addMaterial):
(WebKit::DDModel::RemoteDDMeshProxy::updateMaterial):
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.h:

Canonical link: <a href="https://commits.webkit.org/301613@main">https://commits.webkit.org/301613@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9b0e3d716d8e886db1f6cf68ca4252600e300c13

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126507 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46152 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37068 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133397 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78183 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128378 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46787 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54690 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96266 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/64366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/844bf2d3-cec5-4eac-ae26-ba54cafb04d5) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129455 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37423 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113136 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76739 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e1bc5b7f-c6c8-42b2-9e9e-8f9cf8f407f0) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/125859 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36315 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31313 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76704 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/118557 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107231 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31594 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135944 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/124972 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53199 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40904 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104775 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53685 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109457 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104477 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26653 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49952 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28284 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50601 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53118 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58932 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/158018 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52401 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39538 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55735 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54136 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7d2261bdf1d1a066feab540a2a7acb2a18faa346. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51529

rr-codes reviewed (dismissed):
We must not modularize WebCore before modularizing its dependents (so just JSC, which I have a PR for)

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/fa5d227675ac2e73446a5f5861e8d87be3b48346. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51529

mwyrzykowski commented:
> We must not modularize WebCore before modularizing its dependents (so just JSC, which I have a PR for)

I agree and in any case it shouldn't be part of this PR, I removed the changes to WebCore.xcconfig

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6acede12ec6b85c078a4a27472d873cf0fb5dd85. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51529

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9b161f90a4219f7deb64e3fc9310aff7aac66a97. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51529

etiennesegonzac commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -632,22 +892,437 @@ final class Converter {
         return webRequest
     }
 
+    static func toWebSemantic(_ semantic: DDImageAsset.Semantic) -> WebDDSemantic {
```
I think this can get the `toWebDDDataType` treatment.

etiennesegonzac commented on Source/WebCore/html/canvas/UsdModelLoader.swift:
```diff
@@ -632,22 +892,437 @@ final class Converter {
+        WebDDImageAsset(path: asset.path, utType: asset.utType, data: asset.data, semantic: toWebSemantic(asset.semantic), identifier: id)
+    }
+
+    static func toWebDDNodeType(_ node: DDMaterialGraph.Node) -> WebDDNodeType {
```
Ditto.

webkit-early-warning-system commented:
EWS run on current version of this PR (hash https://github.com/WebKit/WebKit/commit/9f0f0ed52eb3c6a3151ab4600e29b648301c2ac5)<details>

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126441 "Passed style check") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36989 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128312 "Passed tests") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129389 "Passed tests") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/138/builds/36989 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [⏳ 🧪 api-wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | | ⏳ 🛠 vision-apple 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/125793 "Passed tests") | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/138/builds/36989 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/138/builds/36989 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [⏳ 🛠 gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/124906 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled JSC (warnings); Running jscore-test") | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-2-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 gtk-wk2 ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/138/builds/36989 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [⏳ 🧪 api-gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/138/builds/36989 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [⏳ 🛠 playstation ](https://ews-build.webkit.org/#/builders/PlayStation-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-safer-cpp ](https://ews-build.webkit.org/#/builders/macOS-Safer-CPP-Checks-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/157952 "Build is in progress. Recent messages:Printed configuration") | | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/157952 "Build is in progress. Recent messages:Printed configuration") | | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-11-Build-EWS "Waiting in queue, processing has not started yet") | | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-11-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | | </details>

mwyrzykowski committed:
[Model element] Support loading textures in GPU process model element
https://bugs.webkit.org/show_bug.cgi?id=299470
rdar://161267992

Reviewed by Etienne Segonzac.

Send materials and textures from the WCP to to GPU process for rendering.

Remove the WebKit:: types which overlapped with WebCore:: types. Previously
we were converted back and forth from otherwise identical WebKit and WebCore
types which seemed redundant and also inefficient.

Test: model-element/gpup-model-element-textures.html
* LayoutTests/model-element/gpup-model-element-textures-expected.png: Added.
* LayoutTests/model-element/gpup-model-element-textures-expected.txt: Added.
* LayoutTests/model-element/gpup-model-element-textures.html: Added.
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.cpp:
(WebCore::DDModel::DDMeshImpl::addMesh):
(WebCore::DDModel::DDMeshImpl::update):
(WebCore::DDModel::convert):
(WebCore::DDModel::DDMeshImpl::addTexture):
(WebCore::DDModel::DDMeshImpl::updateTexture):
(WebCore::DDModel::convertArray):
(WebCore::DDModel::DDMeshImpl::addMaterial):
(WebCore::DDModel::DDMeshImpl::updateMaterial):
* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.h:
* Source/WebCore/Modules/Model/InternalAPI/DDImageAsset.h: Renamed from Source/WebKit/Shared/Model/DDReplaceVertices.h.
* Source/WebCore/Modules/Model/InternalAPI/DDMaterialDescriptor.h: Added.
* Source/WebCore/Modules/Model/InternalAPI/DDMesh.h:
* Source/WebCore/Modules/Model/InternalAPI/DDMeshDescriptor.h:
* Source/WebCore/Modules/Model/InternalAPI/DDMeshPart.h:
* Source/WebCore/Modules/Model/InternalAPI/DDModel.serialization.in:
* Source/WebCore/Modules/Model/InternalAPI/DDReplaceVertices.h:
* Source/WebCore/Modules/Model/InternalAPI/DDTextureDescriptor.h: Renamed from Source/WebKit/Shared/Model/DDVertexLayout.h.
* Source/WebCore/Modules/Model/InternalAPI/DDUpdateMaterialDescriptor.h: Copied from Source/WebCore/Modules/Model/InternalAPI/DDReplaceVertices.h.
* Source/WebCore/Modules/Model/InternalAPI/DDUpdateMeshDescriptor.h:
* Source/WebCore/Modules/Model/InternalAPI/DDUpdateTextureDescriptor.h: Copied from Source/WebCore/Modules/Model/InternalAPI/DDReplaceVertices.h.
* Source/WebCore/Modules/Model/InternalAPI/DDVertexAttributeFormat.h:
* Source/WebCore/Modules/Model/InternalAPI/DDVertexLayout.h:
* Source/WebCore/Modules/Model/ModelDDInlineConverters.h:
(WebCore::toCpp):
(WebCore::toVector):
* Source/WebCore/Modules/Model/ModelDDTypes.h:
* Source/WebCore/Modules/model-element/DDModelPlayer.h:
* Source/WebCore/Modules/model-element/DDModelPlayer.mm:
(WebCore::DDModelPlayer::load):
(WebCore::DDModelPlayer::update):
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
(): Deleted.
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayer.cpp:
(WebCore::DummyModelPlayer::load):
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayer.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/canvas/UsdModelLoader.swift:
(update(_:)):
(meshCreated(_:)):
(meshUpdated(_:data:)):
(meshDestroyed(_:)):
(materialCreated(_:)):
(materialUpdated(_:data:)):
(materialDestroyed(_:)):
(textureCreated(_:)):
(textureUpdated(_:data:)):
(textureDestroyed(_:)):
(Converter.webUpdateMeshRequestFromUpdateMeshRequest(_:)):
(Converter.toWebSemantic(_:)):
(Converter.toWebImageAsset(_:id:)):
(Converter.toWebDDNodeType(_:)):
(Converter.toWebDDBuiltin(_:)):
(Converter.constantValues(_:WebDDConstant:)):
(Converter.toWebDDConstantContainer(_:)):
(Converter.toWebNodes(_:)):
(Converter.toWebEdges(_:)):
(Converter.toWebDDDataType(_:)):
(Converter.toWebInputOutputs(_:)):
(Converter.toWebPrimvars(_:)):
(Converter.toWebMaterialGraph(_:id:)):
(Converter.webAddMeshRequestFromAddMeshRequest(_:)):
(Converter.webUpdateTextureRequestFromUpdateTextureRequest(_:)):
(Converter.webAddTextureRequestFromAddTextureRequest(_:)):
(Converter.webUpdateMaterialRequestFromUpdateMaterialRequest(_:)):
(Converter.webAddMaterialRequestFromAddMaterialRequest(_:)):
(WebUSDModelLoader.textureAddedCallback):
(WebUSDModelLoader.textureUpdatedCallback):
(WebUSDModelLoader.materialAddedCallback):
(WebUSDModelLoader.materialUpdatedCallback):
(WebUSDModelLoader.retainedRequests):
(WebUSDModelLoader.update(_:)):
(WebUSDModelLoader.requestCompleted(_:)):
(WebUSDModelLoader.updateMesh(_:)):
(WebUSDModelLoader.addTexture(_:)):
(WebUSDModelLoader.updateTexture(_:)):
(WebUSDModelLoader.addMaterial(_:)):
(WebUSDModelLoader.updateMaterial(_:)):
(WebUSDModelLoader.meshAdded(_:)):
* Source/WebGPU/WebGPU/DDMesh.h:
* Source/WebGPU/WebGPU/DDMesh.mm:
(WebGPU::DDMesh::DDMesh):
(WebGPU::convert):
(WebGPU::convertDescriptor):
(WebGPU::DDMesh::update):
(WebGPU::DDMesh::addMesh):
(WebGPU::DDMesh::addTexture):
(WebGPU::DDMesh::updateTexture):
(WebGPU::DDMesh::addMaterial):
(WebGPU::DDMesh::updateMaterial):
(wgpuDDTextureUpdate):
(wgpuDDTextureAdd):
(wgpuDDMaterialUpdate):
(wgpuDDMaterialAdd):
* Source/WebGPU/WebGPU/DDModelTypes.h:
* Source/WebGPU/WebGPU/UsdModelRenderer.swift:
(NodeStore.nodeDefinitionStore):
(Helper.makeDirectTextureDescriptorFromMTLTexture(_:)):
(Helper.pixelFormat):
(sgDataTypeFromInputOutputDataType(_:)):
(nodes):
(allNodes):
(connections):
(customSurfaceShaderParameterBindings):
(assetPathToPropertyName):
(pbrInputNameToType(_:)):
(convert(_:)):
(convertToAddMeshRequest(_:id:)):
(DDBridgeReceiver.meshInstances):
(DDBridgeReceiver.meshResources):
(DDBridgeReceiver.textures):
(DDBridgeReceiver.textureFromPath):
(DDBridgeReceiver.render(with:)):
(DDBridgeReceiver.addMesh(_:identifier:)):
(DDBridgeReceiver.addTextureAsync(_:identifier:)):
(DDBridgeReceiver.addTexture(_:identifier:)):
(DDBridgeReceiver.updateTextureAsync(_:identifier:)):
(DDBridgeReceiver.updateTexture(_:identifier:)):
(DDBridgeReceiver.addMaterialAsync(_:identifier:)):
(DDBridgeReceiver.addMaterial(_:identifier:)):
(DDBridgeReceiver.updateMaterialAsync(_:identifier:)):
(DDBridgeReceiver.updateMaterial(_:identifier:)):
(DDBridgeReceiver.updateMeshAsync(_:identifier:)):
(DDBridgeReceiver.updateMesh(_:identifier:)):
(DDBridgeReceiver.device): Deleted.
(DDBridgeReceiver.context): Deleted.
(DDBridgeReceiver.scene): Deleted.
(DDBridgeReceiver.meshes): Deleted.
(DDBridgeReceiver.camera): Deleted.
(DDBridgeReceiver.commandQueue): Deleted.
(DDBridgeReceiver.captureManager): Deleted.
(DDBridgeReceiver.setDeviceWith(_:)): Deleted.
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.cpp:
(WebKit::RemoteDDMesh::addMesh):
(WebKit::RemoteDDMesh::update):
(WebKit::RemoteDDMesh::addTexture):
(WebKit::RemoteDDMesh::updateTexture):
(WebKit::RemoteDDMesh::addMaterial):
(WebKit::RemoteDDMesh::updateMaterial):
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.h:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/Model/DDMeshDescriptor.cpp: Removed.
* Source/WebKit/Shared/Model/DDMeshDescriptor.h: Removed.
* Source/WebKit/Shared/Model/DDMeshDescriptor.serialization.in: Removed.
* Source/WebKit/Shared/Model/DDMeshPart.cpp: Removed.
* Source/WebKit/Shared/Model/DDMeshPart.h: Removed.
* Source/WebKit/Shared/Model/DDMeshPart.serialization.in: Removed.
* Source/WebKit/Shared/Model/DDReplaceVertices.cpp: Removed.
* Source/WebKit/Shared/Model/DDReplaceVertices.serialization.in: Removed.
* Source/WebKit/Shared/Model/DDUpdateMeshDescriptor.cpp: Removed.
* Source/WebKit/Shared/Model/DDUpdateMeshDescriptor.h: Removed.
* Source/WebKit/Shared/Model/DDUpdateMeshDescriptor.serialization.in: Removed.
* Source/WebKit/Shared/Model/DDVertexAttributeFormat.h: Removed.
* Source/WebKit/Shared/Model/DDVertexAttributeFormat.serialization.in: Removed.
* Source/WebKit/Shared/Model/DDVertexLayout.serialization.in: Removed.
* Source/WebKit/Shared/Model/ModelConvertFromBackingContext.h:
* Source/WebKit/Shared/Model/ModelConvertToBackingContext.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.cpp:
(WebKit::DDModel::RemoteDDMeshProxy::addMesh):
(WebKit::DDModel::RemoteDDMeshProxy::update):
(WebKit::DDModel::RemoteDDMeshProxy::addTexture):
(WebKit::DDModel::RemoteDDMeshProxy::updateTexture):
(WebKit::DDModel::RemoteDDMeshProxy::addMaterial):
(WebKit::DDModel::RemoteDDMeshProxy::updateMaterial):
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.h:

Canonical link: https://commits.webkit.org/301613@main
